From dbc2005c1374c95ff1fbab5338698c3ebecfeeb3 Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Mon, 28 Sep 2020 12:40:08 +0300
Subject: [PATCH] [SW-21029] goya2/goya2_arc: ppboot: clear bss section

This patch clears the bss section such that the global
variables are initialized to 0 before going to the
main function of ppboot

Change-Id: I4c693f0a49a12ca0a184d64a40c4629e296e7e69
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/ppboot/crt0_ppboot.S b/ppboot/crt0_ppboot.S
index b4bd0bb..e4b2e30 100644
--- a/ppboot/crt0_ppboot.S
+++ b/ppboot/crt0_ppboot.S
@@ -16,13 +16,30 @@
 /* lf2flt: to avoid compiler (buildroot wrapper) warning during linkage */
 .global	lf2flt
 lf2flt:
+
+	/* Clear BSS */
+	mov     r5, __bss_start
+	sub     r6, __bss_end, r5
+	lsr.f   lp_count, r6, 2
+	lpnz    1f
+	st.ab   0, [r5, 4]
+1:
 #endif
 
 #ifdef __aarch64__
 	ldr	x0, =__stack_top
 	bic	sp, x0, #0xf	/* 16-byte alignment for ABI compliance */
 	mov	x29, sp		/* Set FP (x29) */
+
+	/* Clear BSS */
+	ldr	x0, =__bss_start
+	ldr	x1, =__bss_end
+clear_loop:
+	str	xzr, [x0], #8
+	cmp	x0, x1
+	b.lo	clear_loop
 #endif
+
 #ifdef __arc__
 	push_s	blink
 #endif
